### PR TITLE
Fix decimal to anydata cast error when using cloneWithType

### LINKED ISSUE #26624: `xxxWithType` doesn't work when a `decimal` value is attempted to be used as `anydata`
**Description:**
$title for `cloneWithType` and `fromJsonWithType`.

**Steps to reproduce:**
```ballerina
import ballerina/io;

type OpenRecord record {

};

public function main() {
    map<json> mp = {
        name: "foo",
        factor: 1.23d
    };

    io:println(mp.cloneWithType(OpenRecord));
}
```

This fails when trying to add the `decimal` field for the rest type (`anydata`).
```cmd
error("{ballerina/lang.typedesc}ConversionError",message="'map<json>' value cannot be converted to 'OpenReco
rd'")
```

Workaround is to explicitly define a `decimal`-typed field.
```ballerina
type OpenRecord record {
    decimal factor;
};
```

or include `decimal` explicitly in the target rest type.
```ballerina
type OpenRecord record {|
    anydata|decimal...;   
|};
```

The is type checks work as expected.

`mp["factor"] is decimal` and `mp["factor"] is anydata` print true.

**Affected Versions:**
slp5

### PR DESCRIPTION
**Pull request by dulvinw**
Fixes #26624 

```
type OpenRecord record {

};

public function main() {
    map<json> mp = {
        name: "foo",
        factor: 1.23d
    };

    var e = mp.cloneWithType(OpenRecord);

    if (e is error) {
       panic("cloneWithTypeFail");
    }
}
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/conversion/native-conversion-stampable-values.bal [hunk 1]
```diff
@@ -66,3 +66,35 @@ function testConvertStampTupleToMap() returns [[string, Employee], [string, Empl
...
+
+function testConvertMapJsonWithDecimalToOpenRecord() {
+    map<json> mp = {
+        name: "foo",
+        factor: 1.23d
+    };
+
+    var or = mp.cloneWithType(OpenRecord);
+
+    if(or is error) {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/expressions/conversion/native-conversion-stampable-values.bal:82]:
```suggestion
    if (or is error) {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/conversion/native-conversion-stampable-values.bal [hunk 2]
```diff
@@ -66,3 +66,35 @@ function testConvertStampTupleToMap() returns [[string, Employee], [string, Empl
...
+}
+
+function assert(anydata actual, anydata expected) {
+    if (expected != actual) {
+        typedesc<anydata> expT = typeof expected;
+        typedesc<anydata> actT = typeof actual;
+        string reason = "expected [" + expected.toString() + "] of type [" + expT.toString()
+                            + "], but found [" + actual.toString() + "] of type [" + actT.toString() + "]";
+        error e = error(reason);
+        panic e;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/expressions/conversion/native-conversion-stampable-values.bal:98]:
```suggestion
        panic error(reason);
```

**Review by MaryamZi — COMMENTED** [review]
Shall we add a test for when decimal is part of a union target (e.g., `decimal|string`) and a negative test too?

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/conversion/native-conversion-stampable-values.bal (hunk 1 above)
Comment (dulvinw) [reply tests/jballerina-unit-test/src/test/resources/test-src/expressions/conversion/native-conversion-stampable-values.bal]: Fixed

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/conversion/native-conversion-stampable-values.bal (hunk 2 above)
Comment (dulvinw) [reply tests/jballerina-unit-test/src/test/resources/test-src/expressions/conversion/native-conversion-stampable-values.bal]: Fixed

**Review by hasithaa — COMMENTED** [review]
Looks good to me

**Review by MaryamZi — APPROVED** [review]
@dulvinw, as disucssed shall we also add some tests for the same with `fromJsonWithType` in a separate PR?

**Pull request merged**